### PR TITLE
Allow throwing of edm::Exceptions from python

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Only a selected set of edm::errors codes are allowed.

#### PR validation:

Code compiles and a simple script shows it works.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: FWCore/PythonParameterSet/src/PyBind11Module.h [hunk 1]
```diff
@@ -39,6 +40,8 @@ PYBIND11_MODULE(libFWCorePythonParameterSet, m) {
         std::rethrow_exception(p);
     } catch (const cms::Exception &e) {
       PyErr_SetString(PyExc_RuntimeError, e.what());
+    } catch (const pybind11::error_already_set &e) {
+      std::cout << "pybind " << e.what() << std::endl;
```
Comment (makortel) [inline FWCore/PythonParameterSet/src/PyBind11Module.h:44]: Is this printout still useful?

File: FWCore/PythonParameterSet/src/PyBind11Wrapper.cc [hunk 2]
```diff
@@ -1,9 +1,29 @@
...
 
   void pythonToCppException(const std::string& iType, const std::string& error) {
+    auto colon = error.find(':');
+    if (colon != std::string::npos) {
+      if (error.substr(0, colon) == "EDMException") {
+        auto errorNameStart = error.find('{');
+        auto errorNameEnd = error.find('}');
+        if (errorNameStart != std::string::npos and errorNameEnd != std::string::npos) {
+          auto errorName = error.substr(errorNameStart + 1, errorNameEnd - errorNameStart - 1);
+          std::cout << "EDMException '" << errorName << "'" << std::endl;
```
Comment (makortel) [inline FWCore/PythonParameterSet/src/PyBind11Wrapper.cc:15]: Can this printout be removed?

File: FWCore/PythonParameterSet/src/PyBind11Wrapper.cc [hunk 3]
```diff
@@ -1,9 +1,29 @@
...
+          if ("Configuration" == errorName) {
+            std::cout << "throwing edm::Exception" << std::endl;
```
Comment (makortel) [inline FWCore/PythonParameterSet/src/PyBind11Wrapper.cc:17]: Can this printout be removed?

File: FWCore/PythonParameterSet/src/PyBind11Wrapper.cc [hunk 4]
```diff
@@ -1,9 +1,29 @@
...
   void pythonToCppException(const std::string& iType, const std::string& error) {
+    auto colon = error.find(':');
+    if (colon != std::string::npos) {
+      if (error.substr(0, colon) == "EDMException") {
+        auto errorNameStart = error.find('{');
+        auto errorNameEnd = error.find('}');
+        if (errorNameStart != std::string::npos and errorNameEnd != std::string::npos) {
+          auto errorName = error.substr(errorNameStart + 1, errorNameEnd - errorNameStart - 1);
+          std::cout << "EDMException '" << errorName << "'" << std::endl;
+          if ("Configuration" == errorName) {
```
Comment (makortel) [inline FWCore/PythonParameterSet/src/PyBind11Wrapper.cc:16]: Would it be feasible to generalize the name-to-error code translation by exploiting the information in
https://github.com/cms-sw/cmssw/blob/52c0dacc1af916e537cb419a39ec5845ae5087bf/FWCore/Utilities/src/EDMException.cc#L9-L10
?

On the other hand, the number of error codes to throw from the python code is likely small (I can think of one to add, but not really more)

File: FWCore/ParameterSet/python/Config.py [hunk 5]
```diff
@@ -27,6 +27,14 @@
 if sys.getrecursionlimit()<5000:
     sys.setrecursionlimit(5000)
 
+class edm(object):
+    class error(object):
+        #Allowed errors to used within Python
+        Configuration = "{Configuration}"
+
+class EDMException(Exception):
+    pass
```
Comment (makortel) [inline FWCore/ParameterSet/python/Config.py:36]: It is not crystal clear to me how these would be used. Based on the implementation I'd guess
```py
raise EDMException(edm.error.Configuration + "Real exception message")
```

How about something along the following interface?
```py
raise EDMException(edm.error.Configuration, "Real exception message")
```

File: FWCore/Framework/bin/cmsRun.cpp [hunk 6]
```diff
@@ -240,6 +240,8 @@ int main(int argc, char* argv[]) {
       try {
         std::unique_ptr<edm::ParameterSet> parameterSet = edm::readConfig(fileName, argc, argv);
         processDesc.reset(new edm::ProcessDesc(std::move(parameterSet)));
+      } catch (edm::Exception const&) {
+        throw;
```
Comment (makortel) [inline FWCore/Framework/bin/cmsRun.cpp:244]: I wonder if adding the calling context to the exception would be helpful to interpret the exception message?

**Review by makortel — COMMENTED** [review]
How about adding a unit test?

File: FWCore/Framework/bin/cmsRun.cpp (hunk 6 above)
Comment (Dr15Jones) [reply FWCore/Framework/bin/cmsRun.cpp:244]: The call actually does have a lot of context. E.g.
```
----- Begin Fatal Exception 14-Dec-2022 14:21:44 CST-----------------------
An exception of category 'Configuration' occurred while
   [0] Processing the python configuration file named test_pyhon_error.py
Exception Message:
test message

At:
  test_pyhon_error.py(11): <module>
----- End Fatal Exception -------------------------------------------------
```

for my trivial test case where the configuration file only has the `raise` call.

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+1

**Comment by perrotta** [comment]
+1

**Pull request merged**